Add SEPA payments

katzmo opened the pull request.

katzmo committed:
Refactor for multiple payment methods.

katzmo committed:
JS: update handleCard* → confirmCard*

handleCardPayment() and handleCardSetup() have been deprecated by Stripe
and should be replaced by confirmCardPayment() and confirmCardSetup().

katzmo committed:
Basic implemention of SEPA payments.

katzmo committed:
SEPA: default to ensure Stripe required fields.

katzmo committed:
JS: set IBAN placeholder country if available.

There is no way to hide the placeholder for the IBAN element. Default
country when not set is 'DE', let’s use the same fallback.

katzmo committed:
SEPA: add authorization text

katzmo committed:
Merge branch 'fix-error-messages' into sepa

katzmo committed:
JS: handle errors for IBAN field.

torotil commented on package/src/method-element.js:
```diff
@@ -146,6 +154,29 @@ class MethodElement {
+    let handler = 'confirm'
+    let data = { payment_method: this.extraData() }
+    if (this.settings.intent_methods.includes('sepa_debit')) {
+      handler += 'SepaDebit'
```
I’d prefer to have the full handler name here (ie. `'confirmSepaDebit'`). Makes grepping easier ;-) … In general I think a mapping from `this.settings.intent_type` to the appropriate `name` and `handler` would be a better choice than relying on some pattern in the stripe API’s method names.

katzmo replied:
[Tried to make it a bit more transparent!](https://github.com/moreonion/stripe_payment/pull/9/commits/6716499c259bbd620182acfdbbe485ef0d597c5d) Having the full handler name is IMHO neither feasable nor easier to read: the first part is defined by the intent method, the second part by the intent type, so each new method would require the same nested if to choose between it’s 2 handlers… (If Stripe.js changes that, we’d have to change our code anyway)

As for the name, that actual *is* the camel case version of the [type passed by php](https://github.com/moreonion/stripe_payment/blob/sepa/src/StripeForm.php#L32), can’t make that more explicit?

torotil replied:
Could you maybe include a link to appropriate page in the stripe docs? I’m having a hard time guessing what the possible values are, and how many different paths the code could go on from there.

katzmo replied:
There is no doc listing *all* handlers, they're all in separate chapters of the doc…

torotil replied:
Ok, then there is nothing in the docs that describes how the handler-name should be created? — If there is nothing in there we shouldn’t assume a pattern.

katzmo replied:
Turns out there are only 5 possible handlers (that I could find in the current doc) anyway: 

* confirmCardPayment
* confirmSepaDebitPayment
* confirmIdealPayment
* confirmCardSetup
* confirmSepaDebitSetup

→ [amended commit](https://github.com/moreonion/stripe_payment/pull/9/commits/9efa3706af482df1108f5178e32ea2f9e7dddf3b)

torotil commented on src/SepaController.php:
```diff
@@ -15,6 +15,22 @@ public function __construct() {
+          'required' => 1,
+        ],
+      ]
+    ];
```
These fields should then also have `$element['#required']` in the method configuration.

katzmo replied:
Ok… what is the "method configuration" if that’s not it? :thinking:

torotil replied:
I meant `CustomerDataForm::fields()`. The element definitions there are used to generate the method configuration form.

katzmo replied:
I don’t really understand? In `CustomerDataForm::configurationForm()`, the `controller_data_defaults` are merged with `CustomerDataForm::fields()`. Why should they be in the elements definitions before that?

torotil replied:
Because a field being actually required by the payment provider (#required) is different from the setting to make it required by choice (the default value).

torotil reviewed (changes requested):
Looks great (except for the comments)!

katzmo committed:
JS: fix function name

katzmo committed:
JS: make handler name more transparent

katzmo committed:
JS: expand SEPA error codes

katzmo committed:
JS: fix clientside validation form reset.

We just want to clear the error messages, resetForm() also removes the
form values.

torotil committed:
Use the elements module to have input type="email" elements.

katzmo committed:
Extend controller required flag.

- Pass required from settings.
- Add to JS as data attribute for webform_paymethod_select.js.

katzmo committed:
Restrict config options for Stripe required fields.

Make sure fields required by Stripe can’t be disabled or set to not
required in the configuration form.

torotil committed:
Merge remote-tracking branch 'origin/7.x-2.x' into sepa
